nvim/api.lua

```diff
@@ -0,0 +1,536 @@
+  return string.format('Expected %s[%s] but we found `%s`(%s)', typ, tostring(typ_of), tostring(value), type(value))
+end
+
+--- Implemeted Nvim Basic types
+local Nil = {
+  type_name = 'Nil',
+  basic_type = true,
+  new = function (v) 
+    if type(v) == 'nil' then return nil end
+    return nil, error_convert('Nil', v)
+  end
+}
+
+local Void = {
+  type_name = 'void',
+  basic_type = true,
+  new = function (v) return nil end
+}
+
+local Boolean = {
+  type_name = 'Boolean',
+  basic_type = true,
+  new = function (v)
+    if type(v) == 'boolean' then
+      return v
+    end
+    return nil, error_convert('Boolean', v)
+  end
+}
+
+local Integer = {
+  type_name = 'Integer',
+  basic_type = true,
+  new = function (v)
+    if type(v) == 'number' then
+      return v
+    end
+    return nil, error_convert('Integer', v)
+  end
+}
+
+local Float = {
+  type_name = 'Float',
+  basic_type = true,
+  new = function (v) 
+    if type(v) == 'number' then
+      return v
+    end
+    return nil, error_convert('Float', v)
+  end
+}
+
+local String = {
+  type_name = 'String',
+  basic_type = true,
+  new = function (v) 
+    if type(v) ~= 'string' then
+      return nil, error_convert('String', v)
+    else
+      return tostring(v)
+    end
+  end
+}
+
+local Array = {
+  type_name = 'Array',
+  basic_type = true,
+  new = function (array, of_type) 
+    if type(array) ~= "table" then 
+      return nil, error_convert('Array', array)
+    end
+    if of_type then
+      local array_of = {}
+      for i, value in ipairs(array) do
+        local typed_value, err = of_type.new(value)
+        if not typed_value then 
+          return nil, error_convert_generic('ArrayOf', of_type, value)
+        end
+        array_of[i] = typed_value
+      end
+      return array_of
+    else
+      return util.table_icopy(array)
+    end
+  end
+}
+
+local Dictionary = {
+  type_name = 'Dictionary',
+  basic_type = true,
+  new = function (dictionary) 
+    if type(dictionary) ~= 'table' then
+      return nil, error_convert('Dictionary', dictionary)
+    end
+    return util.table_kvcopy(dictionary)
+  end
+}
+
+local Object = {
+  type_name = 'Object',
+  basic_type = true,
+  new = function(v) return v end,
+}
+
+--- Table to store the basic types in NVim
+local type_tables = {
+  Nil = Nil,
+  Boolean = Boolean,
+  Integer = Integer,
+  Float = Float,
+  String = String,
+  Array = Array,
+  ArrayOf = util.table_kvcopy(Array, {
+    type_generic = 'Array',
+    regex = '(ArrayOf)%((%w+).*%)' -- TODO: We ignore the # of items in the type 
+  }),
+  void = Void, 
+  Dictionary = Dictionary,
+  Object = Object
+}
+
+--- Implemented Ext NVim types
+-- Unknown ext types will still be registered in the Api but
+-- without a fancy wrapper around them.
+local ext_types_table = {
+  Buffer = Buffer,
+  Window = Window,
+  Tabpage = Tabpage
+}
+
+local err_types_table = {
+}
+
+--- Type Buffer
+Buffer.__index = Buffer
```

Comment:
We can remove them completely but my idea was to make the client to know about some types so we can add even higher methods over them in the future making the library nicer and easier to use.

In theory we only use the `knowledge` about those concrete types if we can discover them from in the API discovery phase so if Nvim does not provide them then the client will ignore those concrete types and we  can still use the types provided by Nvim but without having fancy methods on them (there are some unit tests proving that). Basically we achieve that functionality having two kind of types: `basic types` and `extended types`, if we know something about the extended types (the concrete types you talk about in the code) we use that info with the extended type discovered otherwise we wrap them in a generic plain `Object` class that only adds functionality to call the methods coming from the API (everything dynamic at this point).

Answering your question, we could have problems if we add a fancy method on any concrete type relying on a low level method discovered from the API that was removed. We can add some checks to match the required low level methods and their signatures before adding them giving a nice error in case they dont match (keep in mind that the low level functions provided by Nvim will be available always using the client).
In any case if this happens it's easy to fix it supporting the new changes since it's well known where the changes in the client are required.

I hope i expressed myself with clarity.